### PR TITLE
Flag re-implementation

### PR DESCRIPTION
**Pull request by jszwedko**
This PR is a WIP of rethinking the abstraction around flags

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -223,14 +229,15 @@ func (a *App) RunAsSubcommand(ctx *Context) (err error) {
 	// append help to commands
 	if len(a.Commands) > 0 {
 		if a.Command(helpCommand.Name) == nil && !a.HideHelp {
-			a.Commands = append(a.Commands, helpCommand)
-			if (HelpFlag != BoolFlag{}) {
+			a.appendCommand(helpCommand)
+
+			if HelpFlag != (&BoolFlag{}) {
```
Comment (jszwedko) [inline app.go]: I think you want `*HelpFlag != BoolFlag{}` here lest it compare pointer addresses.

Comment (meatballhat) [inline app.go]: 🙀 I'll figure this out, but:

```
$ go get -x ./...
WORK=/var/folders/1y/6fv14q917p78px9mvr7602ph0000gp/T/go-build063161737
mkdir -p $WORK/github.com/codegangsta/cli/_obj/
mkdir -p $WORK/github.com/codegangsta/
cd /Users/me/gopath/src/github.com/codegangsta/cli
/Users/me/.gimme/versions/go1.6.2.darwin.amd64/pkg/tool/darwin_amd64/compile -o $WORK/github.com/codegangsta/cli.a -trimpath $WORK -p github.com/codegangsta/cli -complete -buildid 5c46cf0b631dcb00d49d9dac77d4038bdbd33b5f -D _/Users/me/gopath/src/github.com/codegangsta/cli -I $WORK -pack ./app.go ./category.go ./cli.go ./command.go ./context.go ./errors.go ./flag.go ./flag_set.go ./funcs.go ./help.go
# github.com/codegangsta/cli
./app.go:125: invalid operation: *HelpFlag != BoolFlag literal (struct containing []string cannot be compared)
./app.go:234: invalid operation: *HelpFlag != BoolFlag literal (struct containing []string cannot be compared)
./command.go:68: invalid operation: *HelpFlag != BoolFlag literal (struct containing []string cannot be compared)
```

Comment (meatballhat) [inline app.go]: :facepalm: I switch to a pointer to flag, so comparing to `nil` is a thing...

Comment (jszwedko) [inline app.go]: Doh, of course!

**Comment by meatballhat** [comment]
I'm thinking I'd like to take this in a different direction, so closing out.  Please yell if you'd like any part of this repackaged and PR'd.

**Pull request closed without merging**